File: learndesignpatterns/strategy-go/strategy.go
package main

import "fmt"

type FlyBehavior interface {
	fly()
}

type FlyWithWing struct {
}

func (fw *FlyWithWing) fly() {
	fmt.Println("FlyWithWing")
}

type FlyNoWay struct {
}

func (fnw *FlyNoWay) fly() {
	fmt.Println("FlyNoWay")
}

type QuackBehavior interface {
	quack()
}

type Quack struct {
}

func (q *Quack) quack() {
	fmt.Println("quack")
}

type Squeak struct {
}

func (s *Squeak) quack() {
	fmt.Println("squeak")
}

type Duck struct {
	flyBehavior   FlyBehavior
	quackBehavior QuackBehavior
}

func (d *Duck) setFlyBehavior(flyBehavior FlyBehavior) {
	d.flyBehavior = flyBehavior
}

func (d *Duck) setQuackBehavior(quackBehavior QuackBehavior) {
	d.quackBehavior = quackBehavior
}

func (d *Duck) performFly() {
	d.flyBehavior.fly()
}

func (d *Duck) performQuack() {
	d.quackBehavior.quack()
}

func (d *Duck) display() {
	fmt.Println("Duck display")
}

func main() {
	pd := new(Duck)
	pFly := new(FlyWithWing)
	pd.setFlyBehavior(pFly)
	pSqueak := new(Squeak)
	pd.setQuackBehavior(pSqueak)
	pd.display()
	pd.performFly()
	pd.performQuack()
}

File: learndesignpatterns/strategy-cc/strategy.cc
#include "strategy.h"
#include <iostream>

void FlyWithWing::fly()
{
	std::cout<<"fly with wing"<<std::endl;
}

void FlyNoWay::fly()
{
	std::cout<<"fly no way"<<std::endl;
}

void Quack::quack()
{
	std::cout<<"quack"<<std::endl;
}

void Squeak::quack()
{
	std::cout<<"squeak"<<std::endl;
}

Duck::Duck(): pflyBehavior(NULL), pquackBehavior(NULL)
{
}

Duck::~Duck()
{
	if(pflyBehavior){
		delete pflyBehavior;
		pflyBehavior = NULL;
	}

	if(pquackBehavior){
		delete pquackBehavior;
		pquackBehavior = NULL;
	}
}

void Duck::swim()
{
	std::cout<<"all duck can swim"<<std::endl;
}

void Duck::display()
{
	std::cout<<"duck display"<<std::endl;
}

void Duck::performQuack()
{
	pquackBehavior->quack();
}

void Duck::performFly()
{
	pflyBehavior->fly();
}

void Duck::setFlyBehavior(FlyBehavior *pflyBehav)
{
	pflyBehavior = pflyBehav;
}

void Duck::setQuackBehavior(QuackBehavior *pquackBehav)
{
	pquackBehavior = pquackBehav;
}

MallardDuck::MallardDuck()
{
	pflyBehavior = new FlyWithWing();
	pquackBehavior = new Quack();
}

MallardDuck::~MallardDuck()
{
	if(pflyBehavior){
		delete pflyBehavior;
		pflyBehavior = NULL;
	}

	if(pquackBehavior){
		delete pquackBehavior;	
		pquackBehavior = NULL;
	}
}

void MallardDuck::display()
{
	std::cout<<"mallard duck display"<<std::endl;
}

RubberDuck::RubberDuck()
{
	pflyBehavior = new FlyNoWay();
	pquackBehavior = new Squeak();
}

RubberDuck::~RubberDuck()
{
	if(pflyBehavior){
		delete pflyBehavior;
		pflyBehavior = NULL;
	}

	if(pquackBehavior){
		delete pquackBehavior;	
		pquackBehavior = NULL;
	}
	
}

void RubberDuck::display()
{
	std::cout<<"rubber duck display"<<std::endl;
}

DecoyDuck::DecoyDuck()
{
	pflyBehavior = new FlyWithWing();
	pquackBehavior = new Quack(); 
}

DecoyDuck::~DecoyDuck()
{
	if(pflyBehavior){
		delete pflyBehavior;
		pflyBehavior = NULL;
	}

	if(pquackBehavior){
		delete pquackBehavior;	
		pquackBehavior = NULL;
	}
}
void DecoyDuck::display()
{
	std::cout<<"decoy duck display"<<std::endl;
}

